A beginner's guide to installing and setting up the latest version of Nginx (#542)
rocky-linux/documentation: docs/guides/web/nginx-mainline.md

* Built-in file caching for speed
* WebSockets
* FastCGI support
* And, of course, IPv6

It’s great! So just `sudo dnf install nginx`, right? Well, not exactly. **Rocky Linux repositories don’t have the latest production-ready version of Nginx.** Since our goal is bug-for bug compatibility with Red Hat Enterprise Linux, you can always ask them to update their repos. Or asking the *Nginx* people might work better (you’ll see what I mean).

What *you* can do, right now, is install the “mainline” branch of Nginx yourself. It has all the latest updates and toys, and (to my mind) a simpler directory structure for its configuration files. Here’s how to see it all for yourself:

!!! Note 

    There's another branch called "stable", but it's actually a little outdated for most use cases. The "mainline" branch is considered by the Nginx developers to be thoroughly-tested and stable enough for everyone to use.

## Prerequisites and Assumptions

You’ll need:

* An internet-connected Rocky Linux machine or server.
* A basic familiarity with the command line.
* The ability to run commands as root, either as the root user or with `sudo`.

> * A text editor of your choice, whether graphical or command-line based. For this tutorial, I’m using *nano*.

Reviewer: ```suggestion
* A text editor of your choice, whether graphical or command-line based. For this tutorial, I’m using `nano`.
```

## Installing the Repository

> This part isn’t quite as simple as installing an extra repository usually is. We’re going to have to create a custom repo file for *dnf* to use, and download *Nginx* from. Technically, we’re sort of repurposing  repositories for CentOS that were made and hosted by *Nginx* themselves. This solution may or may not be viable in the long term, but it’s working great for now.

Reviewer: ```suggestion
This part isn’t quite as simple as installing an extra repository usually is. We’re going to have to create a custom repo file for `dnf` to use, and download *Nginx* from. Technically, we’re sort of repurposing  repositories for CentOS that were made and hosted by *Nginx* themselves. This solution may or may not be viable in the long term, but it’s working great for now.
```

Reviewer 2: I was just thinking about this: There is a module for nginx on Rocky for up-to version 1.20. Would this be better than a separate repository?

Author: I know I mentioned this to you on Mattermost @sspencerwire, but here's the reference for posterity. The Nginx module is the current "stable" branch, and it's fine, but the makers of Nginx say this (in an admittedly older blog post):

"In NGINX nomenclature, 'stable' means that no new features are added (the feature set is stable). Only major bug fixes are committed to that version.

"We develop new features and all bug fixes in the mainline branch. We operate a time‑based release process, so you can expect to see new mainline releases approximately once per month, with exceptional releases when necessary. Over the last year, the mainline has seen the introduction of SPDY 3.1 support, authentication via subrequests, SSL session ticket support, IPv6 support for DNS, and PROXY protocol support. We integrated community contributions for SSL support for uwsgi. We’ve extended error logging, added cache revalidation directives, added SMTP pipelining, added buffering options for FastCGI, improved support for MP4 streaming, and extended handling of byte‑range requests for streaming and caching.

"Note that stable does not mean more reliable or more bug‑free. In fact, the mainline is generally regarded as more reliable because we port all bug fixes to it, and not just critical fixes as for the stable branch. On the other hand, changes in the stable branch are very unlikely to affect third‑party modules. We don’t make the same commitment concerning the mainline, where new features can affect the operation of third‑party modules."

And:

"We recommend that in general you deploy the NGINX mainline branch at all times. The main reason to use the stable branch is that you are concerned about possible impacts of new features, such as incompatibility with third‑party modules or the inadvertent introduction of bugs in new features."

- https://www.nginx.com/blog/nginx-1-6-1-7-released/

I'm thinking that I'll explain this further in the guide, and show how to install from the module as well for people.

Reviewer: For my part, I am still not familiar with the module system to be activated in dnf. But this is the official way.
In this case, I think there can be the "official" version and then an explanation "If you want to have an even newer version"... and keep the external repository and the actual explanation which is fine.
But the officially preferred version by redhat first.
That's good, it adds information to the page.

Translated with www.DeepL.com/Translator (free version)

Author: Well there's Red Hat official, and there's "getting all the latest security patches" official. I lean toward the latter, and I think I'll be saying as much in the guide.

Reviewer 2: @EzequielBruni I'm fine with whatever you feel here. Rocky being a bug-for-bug compatible version of Red Hat means that the code is always going to lag a bit. It's by design. Doing things with your repository is absolutely fine. We do this all the time when we add in third-party repositories to get newer versions of software. Best example here is PHP 8.x. So you can add in the module install as an option, or not. Totally up to you.

Author: Oh I think it's genuinely a good idea to have both methods outlined for the sake of clarity and education. I like being thorough where I can.

And to think this all started because I realized we would need a full Nginx install guide to refer to for the LXD guide I'm writing.

Kind of curious to see what Antoine will have to say about that one. :)

First, make sure your machine is updated:

```bash
sudo dnf update
```

> Now, make sure the *dnf-utils* package is installed, and any command-line text editor you may want:

Reviewer: ```suggestion
Now, make sure the `dnf-utils` package is installed, and any command-line text editor you may want:
```

```bash
sudo dnf install dnf-utils
```

> Once that’s all installed, power up your text editor of choice. You’ll want to create a file called (for the sake of simplicity) “nginx.repo”, and put it in `/etc/yum.repos.d/`. You can do this real quick like so:

Reviewer: ```suggestion
Once that’s all installed, power up your text editor of choice. You’ll want to create a file called (for the sake of simplicity) `nginx.repo`, and put it in `/etc/yum.repos.d/`. You can do this real quick like so:
```

```bash
sudo nano /etc/yum.repos.d/nginx.repo
```

In that file, paste this bit of code, unmodified:

```bash
[nginx-stable]
name=nginx stable repo
baseurl=http://nginx.org/packages/centos/$releasever/$basearch/
gpgcheck=1
enabled=1
gpgkey=https://nginx.org/keys/nginx_signing.key
module_hotfixes=true

[nginx-mainline]
name=nginx mainline repo
baseurl=http://nginx.org/packages/mainline/centos/$releasever/$basearch/
gpgcheck=1
enabled=0
gpgkey=https://nginx.org/keys/nginx_signing.key
module_hotfixes=true
```

This code basically just lets you use the *Nginx*-made-and-hosted repositories for CentOS, and it allows you use the previously-mentioned “stable” branch if you want to. I mean, don’t. But you could.

> Save the file with Control-S (if using *nano*) and exit with Control-X.

Reviewer: ```suggestion
Save the file with <kbd>Control</kbd>-<kbd>S</kbd> (if using `nano`) and exit with <kbd>Control</kbd>-<kbd>X</kbd>.
```

## Installing and Running Nginx

Now, let’s enable the repository file you just made with one simple command:

```bash
sudo yum-config-manager --enable nginx-mainline
```

> Then, install *Nginx*:

Reviewer: ```suggestion
Then, install the package `nginx` from the previously added repository:
```

```bash
sudo dnf install nginx
```

> The terminal will ask you if you’re fine with installing the repository’s GPG key. You need that, so choose “Y” for yes.

Reviewer: ```suggestion
The terminal will ask you if you’re fine with installing the repository’s GPG key. You need that, so choose `Y` for yes.
```

> Once the installation is done, start *Nginx* and enable it to automatically start on reboot all in one go with:

Reviewer: ```suggestion
Once the installation is done, start the `nginx` service and enable it to automatically start on reboot all in one go with:
```

```bash
sudo systemctl enable --now nginx
```

To verify that the lastest version of *Nginx* has been installed, run:

```bash
nginx -v
```

From there, you could just start dropping HTML files into the `/usr/share/nginx/html/` directory to build a simple, static website. The configuration file for the default website/virtual host is called “default.conf” and it’s in `/etc/nginx/conf.d/`.

## Configuring the Firewall

!!! Note

    If you are installing Nginx on a container such as LXD/LXC or Docker, you can just skip this part for now. The firewall should be handled by the host OS.

If you try to view a web page at your machine’s IP address or domain name from another computer, you’re probably going to get a big fat nothing. Well, that’ll be the case as long as you have a firewall up and running.

> Here’s how to open up the necessary ports to actually see your web pages with *firewalld*, Rocky Linux’s default firewall with the `firewall-cmd` command. To add a new port, just run this:

Reviewer: ```suggestion
Here’s how to open up the necessary ports to actually see your web pages with `firewalld`, Rocky Linux’s default firewall with the `firewall-cmd` command. To add a new port, just run this:
```

```bash

> sudo firewall-cmd --permanent --zone=public --add-port=80/tcp

Reviewer: Why not using the service name? `firewall-cmd --add-service=http`
It's easier to understant.
```suggestion
sudo firewall-cmd --permanent --zone=public --add-service=http
```

Author: For the service suggestion, I think I'll mention both. To be fair to me, I lifted this section from my upcoming guide on LXD web servers, and I'm pretty sure the ports matter there.

Reviewer: I think this is a rule that should be set at the editorial charter level. I propose that when the service is standard, we should use the name of the service, when the service is custom, we should use the port.

Reviewer 2: Neither of you are wrong here. They are both (port or service) fundamentally the same. I want someone to show me how I can fine-tune firewalls down to a specific IP address allow rule. If they can show me this, and it's not a complete cluster, then I'll actually move away from iptables to firewalld.

```

Let’s break this down: 

* The `-–permanent` flag tells the firewall to make sure this configuration is used every time the firewall is restarted, and when the server itself is restarted. 
* `–-zone=public` tells the firewall to take incoming connections to this port from everyone.
* Lastly, `–-add-port=80/tcp` tells the firewall to accept incoming connections over port 80, as long as they’re using the Transmission Control Protocol, which is what you want in this case.

To repeat the process for SSL/HTTPS traffic, just run the command again, and change the number.

```bash

> sudo firewall-cmd --permanent --zone=public --add-port=443/tcp

Reviewer: ```suggestion
sudo firewall-cmd --permanent --zone=public --add-service=https
```

```

> These configurations won’t take effect until you force the issue. To do that, tell *firewalld* to relead its configurations, like so:

Reviewer: ```suggestion
These configurations won’t take effect until you force the issue. To do that, tell `firewalld` to relead its configurations, like so:
```

```bash
sudo firewall-cmd --reload
```

> Now, there’s a very small chance that this won’t work. In those rare cases, make *firewalld* do your bidding with the old turn-it-off-and-turn-it-on-again.

Reviewer: No it should not fail...
Maybe this is a remarque, so we could add a Note admonition

> Now, there’s a very small chance that this won’t work. In those rare cases, make *firewalld* do your bidding with the old turn-it-off-and-turn-it-on-again.
> 
> ```bash
> systemctl restart firewalld
> ```

Reviewer: ```suggestion
!!! Note

    Now, there’s a very small chance that this won’t work. In those rare cases, make `firewalld` do your bidding with the old turn-it-off-and-turn-it-on-again.

    ```bash
    systemctl restart firewalld
    ```
```

To make sure the ports have been added properly, run `firewall-cmd --list-all`. A properly-configured firewall will look a bit like this (I have a few extra ports open on my local server, ignore them):

```bash
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: enp9s0
  sources:

>   services: cockpit dhcpv6-client ssh
>   ports: 81/tcp 444/tcp 15151/tcp 80/tcp 443/tcp

Reviewer: If you choose to keep the service:

```suggestion
  services: cockpit dhcpv6-client ssh http https
  ports: 81/tcp 444/tcp 15151/tcp
```

  protocols:
  forward: no
  masquerade: no
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:
```

And that should be everything you need, firewall-wise.

*Now* you should be able to see a web page that looks something like this:

![The Nginx welcome page](nginx/images/welcome-nginx.png)

It’s not much at all, but it means the server is working.

> 

Reviewer: ```suggestion
You can also test that your web page is accessible with the command line :

```
curl -I http://myip/
```

```

## Creating a Server User and Changing the Website Root Folder

While you *can* just drop your website into the default directory and go (and this might be fine for *Nginx* when it’s running inside a container, or on a test/development server), it’s not what we call best practice. Instead, it’s a good idea to create a specific Linux user on your system for your website, and put your website files in a directory made just for that user. 

If you want to build multiple websites, it’s actually a good idea to create multiple users and root directories, both for the sake of organization and the sake of security. 

> In this guide, I’m going to have just the one user: a handsome devil named “www”. We’re going to put all of his website files under a directory in its home folder: `/home/www/`. You can actually put the folder anywhere you want, but on a dedicated server machine, using home folders makes perfect sense.

Reviewer: Not at all!
And by the way selinux will prevent you from doing that.

Author: selinux would prevent which part of those three paragraphs, specifically? I'm running my tests in containers that don't have it.

Reviewer: Is selinux in enforcing mode in your containers ?
SELinux will prevent any docroot that has not httpd_sys_content label (from my memory), and that not the case for /home. You will have to enable a special boolean to allow the docroot in /home.

Author: selinux isn't installed at all in the containers. In fact, a lot of the guides I've tried have suggested /home/. If it's different for Rocky Linux (some of those guides *were* for Alpine), I can change it to somewhere in /usr/share maybe?

Reviewer 2: I understand what @EzequielBruni is doing here. The minimal container installs do not have SELinux. I'm sure it can be installed, but the whole point of those container installs is that they are by design, secured by the host OS.

### Creating the User

First, we make the folder we’re going to use:

```bash
sudo mkdir /home/www/
```

Then, we create the user:

```bash

> sudo adduser -g 'Nginx www user' -h /home/www/ www

Reviewer: This user should be a system user and not having a bash (/bin/nologin for example).

Author: Hmmm. I was taking use cases like client hosting accounts into consideration. Also, is there a specific change to the command to make it a system user?

Maybe we can talk about both use cases.

Reviewer: I think most of the time, ftp or sftp user to access the docroot are virtual (or should be).
We can keep the id of a user creation, even not a system one, but the discourse must be clear on security. It's a good way to do it, but it's not the most secure way.

Author: Okay, all good stuff to know.

Reviewer 2: @EzequielBruni if you read over the vsftp guide that I wrote, this uses virtual users. Pretty cool!

```

That command tells the machine to

* Make a user called “www” (as per the last bit of text), 
* put all of its files in `/home/www`,
* and add it to the following groups: “Nginx”, “www”, and “user”.

> All three are important, but that “Nginx” group does some real magic. It allows the web server to read and modify files that belong to the “www” user, and the “www” user group. See the Rocky Linux [guide to user management](../../books/admin_guide/06-users.md) for more information.

Reviewer: Ah ah ! Good to see that sort of link!
Thx

> You will, at this point, be prompted to give the new user a password. Type it in, press Enter, and repeat.

Reviewer: Don't give a password to a system user if you can.

Author: See above.

From now on, when you’re actually going to add files to your website, it’s a good idea to do it as the web server user. You can log in to the server user account with by running the following command, and then typing in that password you chose:

```bash
sudo su www
```

### Changing the Server Root Folder

Now that you have your fancy new user account, it’s time to make *Nginx* look for your website files in that folder. It;s time to grab your favorite text editor again.

For now, just run:

```bash
sudo nano /etc/nginx/conf.d
```

When the file is open, look for the line that looks like `root   /usr/share/nginx/html;`. Change it to your chosen website root folder, eg. `root   /home/www;`. Save and close the file, then test your *Nginx* configuration to make sure you didn’t skip a semi-colon or anything:

```bash

> nginx -t

Reviewer: Great!

```

If you get the collowing success message, everything went right:

```
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful
```

Then, give the server a soft restart with:

```bash
sudo systemctl reload nginx
```

> If the soft restart doesn’t work, give *Nginx* a kick in the pants with:

Reviewer: Maybe an admonition?

```bash
sudo systemctl restart nginx
```

Any HTML files in your new root folder should now be browsable from… your browser.

### Changing File Permissions

To make sure that *Nginx* can read, write to, and execute any files in the website directory, permissions need to be set properly, especially if you uploaded the files while using the root account. 

First, make sure that all files in the root folder are owned by the server user and its user group with:

```bash
sudo chown -R www:www /home/www
```

And then, to make sure that users who want to actually browse your website can actually see the pages, you should make you can run this command:

```bash

> sudo chmod -R 755 /home/www 

Reviewer: No! Never do that! My god!
You set as executable all your files! It's dangerous!

You should do something like that:
```
find /home/www -type d -exec chmod 755 "{}" \;
```
It that case, only directory are modified

Author: Hmm. I know I only mentioned static sites, but I was also thinking about stuff like PHP apps... but even those usually only require the files in certain directories to be executable. I'll go into a little more detail and include your suggestion.

Reviewer: No no! PHP files don't need to be executable at all. It's a common mistake. Just to be readable is enough.
No executable files inside a docroot please 🙏.
My rules on a production webserver (for example):
chown -R www:www /var/www/html
find /var/www/html -type d -exec chmod 555 "{}" \;
find /var/www/html -type f -exec chmod 444 "{}" \;
chmod 750 /var/www/html/cache/
chmod 640 /var/www/html/cache/*
(if no directory in cache of course or you will need to use find)

Author: Weird. Some apps I've used won't work at all without more permissive rules. but we can recommend what you say as the default. The help files for individual PHP apps can say differently if necessary.

```

That basically gives everyone the right to look at files on the server, but not modify them. Only the owners get to do that. If you’re feeling paranoid, you can run that particular command every time you upload new HTML/CSS/JS/image files to a static website.

> 

Reviewer: For a better security, only files or directory that you want to be writeable should belongs to the dedicated user. For the rest of the files, readonly (o=r) is sufficient to be served by the webserver.

Author: Fair enough. Time to explain more again, I guess.
